pkg/controllers/user/monitoring/config_refresh_handler.go

```diff
@@ -0,0 +1,137 @@
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/client-go/tools/cache"
+)
+
+const (
+	projectNSAnnotation        = "project.cattle.io/namespaces"
+	nsByProjectIndex           = "monitoring.cluster.cattle.io/ns-by-project"
+	promByMemberNamespaceIndex = "monitoring.cluster.cattle.io/prom-by-member-ns"
+)
+
+type ConfigRefreshHandler struct {
+	prometheusClient  rmonitoringv1.PrometheusInterface
+	prometheusIndexer cache.Indexer
+	nsIndexer         cache.Indexer
+}
+
+func (h *ConfigRefreshHandler) syncNamespace(key string, obj *corev1.Namespace) (runtime.Object, error) {
+	if obj == nil || obj.DeletionTimestamp != nil {
+		return obj, nil
+	}
+	promList, err := h.prometheusIndexer.ByIndex(promByMemberNamespaceIndex, obj.Name)
+	if err != nil {
+		return obj, err
+	}
+
+	for _, object := range promList {
+		prometheus, ok := object.(*monitoringv1.Prometheus)
+		if !ok {
+			continue
+		}
+		h.prometheusClient.Controller().Enqueue(prometheus.Namespace, prometheus.Name)
+	}
+
+	return obj, nil
+}
+
+func (h *ConfigRefreshHandler) syncPrometheus(key string, obj *monitoringv1.Prometheus) (runtime.Object, error) {
+	if obj == nil || obj.DeletionTimestamp != nil {
+		return obj, nil
+	}
+
+	nsObject, exists, err := h.nsIndexer.GetByKey(obj.Namespace)
+	if !exists {
+		return nil, apierrors.NewNotFound(schema.GroupResource{
+			Group:    apiv1.NamespaceGroupVersionKind.Group,
+			Resource: "namespace",
+		}, key)
+	}
+	if err != nil {
+		return obj, err
+	}
+	ns := nsObject.(*corev1.Namespace)
+
+	if len(ns.Labels) == 0 {
+		return obj, nil
+	}
+	projectID, ok := ns.Labels[nslabels.ProjectIDFieldLabel]
+	if !ok {
+		return obj, nil
+	}
+
+	namespaces, err := h.getProjectNamespaces(projectID)
+	if err != nil {
+		return obj, err
+	}
+
+	annotationNamespaces, err := getAnnotationNamespaces(obj)
+	if err != nil {
+		return obj, err
+	}
+
+	if !reflect.DeepEqual(namespaces, annotationNamespaces) {
+		newObj := obj.DeepCopy()
+		if newObj.Annotations == nil {
+			newObj.Annotations = make(map[string]string)
+		}
+		data, err := json.Marshal(namespaces)
+		if err != nil {
+			return obj, err
+		}
+
+		newObj.Annotations[projectNSAnnotation] = string(data)
+		return h.prometheusClient.Update(newObj)
+	}
+
+	return obj, nil
+}
+
+func (h *ConfigRefreshHandler) getProjectNamespaces(projectID string) ([]string, error) {
+	nsList, err := h.nsIndexer.ByIndex(nsByProjectIndex, projectID)
+	if err != nil {
+		return nil, err
+	}
+	var rtn []string
+	defer sort.Strings(rtn)
+	for _, object := range nsList {
+		ns, ok := object.(*corev1.Namespace)
+		if !ok {
+			continue
+		}
+		rtn = append(rtn, ns.Name)
+	}
+	return rtn, nil
+}
+
+func getAnnotationNamespaces(obj *monitoringv1.Prometheus) ([]string, error) {
+	var rtn []string
+	defer sort.Strings(rtn)
+	data, ok := obj.Annotations[projectNSAnnotation]
+	if !ok {
+		return rtn, nil
+	}
+	return rtn, json.Unmarshal([]byte(data), &rtn)
```

Comment:
combining this into a single line makes it very hard to read and understand. can you do the unmarshal on a separate line?